app/uk/gov/hmrc/cgtpropertydisposalsfrontend/controllers/returns/uploadsupportingdocs/UploadSupportingEvidenceController.scala

```diff
@@ -0,0 +1,1012 @@
+      upscanService.initiate(DraftReturnId(draftReturnId.toString), cgtRef, LocalDateTime.now()).value.map {
+        case Left(error) =>
+          error.value match {
+            case Left(_) => errorHandler.errorResult(None)
+            case Right(MaxUploads) =>
+              Redirect(routes.UploadSupportingEvidenceController.checkYourAnswers())
+            case Right(_) => errorHandler.errorResult(None)
+          }
+        case Right(upscanInitiateResponse) =>
+          upscanInitiateResponse match {
+            case UpscanInitiateSuccess(reference) =>
+              Ok(
+                uploadSupportingEvidencePage(
+                  uploadEvidenceForm,
+                  reference,
+                  routes.UploadSupportingEvidenceController.doYouWantToUploadSupportingDocuments()
+                )
+              )
+          }
+      }
+    }
+  }
+
+  @SuppressWarnings(Array("org.wartremover.warts.Var", "org.wartremover.warts.Any"))
+  def uploadSupportingEvidenceSubmit(): Action[MultipartFormData[Files.TemporaryFile]] =
+    authenticatedActionWithSessionData(parse.multipartFormData(maxFileSize)).async { implicit request =>
+      val multipart: MultipartFormData[Files.TemporaryFile] = request.body
+      withUploadSupportingEvidenceAnswers(request) { (draftReturnId, _, _, fillingOutReturn, answers) =>
+        multipart
+          .file("file")
+          .map(supportingEvidence =>
+            if (supportingEvidence.filename === "") {
+              multipart.asFormUrlEncoded.get("reference") match {
+                case Some(reference) => {
+                  Future.successful(
+                    BadRequest(
+                      uploadSupportingEvidencePage(
+                        uploadEvidenceForm.bindFromRequest(request.body.asFormUrlEncoded),
+                        UpscanInitiateReference(
+                          reference.headOption.getOrElse("upscan reference not found")
+                        ),
+                        routes.UploadSupportingEvidenceController.doYouWantToUploadSupportingDocuments()
+                      )
+                    )
+                  )
+                }
+                case None => Future.successful(errorHandler.errorResult(None))
+              }
+            } else {
+              val result = for {
+                reference <- EitherT.fromOption(
+                              multipart.dataParts.get("reference").flatMap(_.headOption),
+                              Error("missing upscan file descriptor id")
+                            )
+                maybeUpscanFileDescriptor <- upscanService
+                                              .getUpscanFileDescriptor(
+                                                DraftReturnId(draftReturnId.toString),
+                                                UpscanInitiateReference(reference)
+                                              )
+                                              .leftMap(e => Error("error1"))
```

Comment:
you might as well put useful error messages here - it's not a lot of effort to do. Though do you even need to `leftMap`? Isn't the left type of the `EitherT` already `Error`?

Reply:
Yes. These got missed. I had a nasty bug to fix and I stuck these in to locate the source of it. Will remove these now.

Reply:
removed